### PR TITLE
Add ORIGINAL_TRIGGER link type to ActT/TCT/TSS

### LINKED ISSUE #246: Better linking from ActT to SCC/SCS
### Description
Imagine an activity that is triggered based on an SCM commit or merge. An ActT event is sent with CAUSE set to that SCC or SCS event and triggers.type set to EIFFEL_EVENT. Then if that activity was retriggered manually resulting another AcT event being sent the ActT would have triggers.type set to MANUAL. But how should the previously sent SCC/SCS event be linked? Linking it with CAUSE does not feel right. Another link type should probably be added there, or?

### Motivation
Make the links to SCC/SCS events from ActT events more elaborated

### Exemplification
As described in description

### Benefits
See above

### Possible Drawbacks
None

### PR DESCRIPTION
**Pull request by magnusbaeck**
### Applicable Issues
Fixes #246 

### Description of the Change
Add a new ORIGINAL_TRIGGER link type to ActT, TCT, and TSS to be able to describe the retry/retrigger/rebuild relationship of subsequent activities and the original activity.

### Alternate Designs
We've quibbled over the name of the link type (see linked issue) but implementation-wise there aren't that many options.

### Benefits
Retries can be modeled with Eiffel.

### Possible Drawbacks
Introducing another relationship for activities could make it harder to understand.

### Sign-off

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I    have the right to submit it under the open source license    indicated in the file; or

(b) The contribution is based upon previous work that, to the best    of my knowledge, is covered under an appropriate open source    license and I have the right under that license to submit that    work with modifications, whether created in whole or in part    by me, under the same open source license (unless I am    permitted to submit under a different license), as indicated    in the file; or

(c) The contribution was provided directly to me by some other    person who certified (a), (b) or (c) and I have not modified    it.

(d) I understand and agree that this project and the contribution    are public and that a record of the contribution (including all    personal information I submit with it, including my sign-off) is    maintained indefinitely and may be redistributed consistent with    this project or the open source license(s) involved.

Signed-off-by: Magnus Bäck \<magnus.back@axis.com>

### CODE REVIEW

**Comment by magnusbaeck** [comment]
Note that this PR branch is based on the #354 PR branch. You'll want to [filter the PR diff view to only show the last commit](https://github.com/eiffel-community/eiffel/pull/355/commits/8a5246e3713b9d637fd2b6d17bd898bbab833f2c). I deliberately haven't added new versions of TCT and TSS since you might have opinions about the wording of the ORIGINAL_TRIGGER documentation.

**Comment by magnusbaeck** [comment]
The PR branch has been rebased and it now based on the master branch instead of the (now merged) PR.

**Comment by magnusbaeck** [comment]
Here's a diff between newly added event versions and their predecessors:

```
diff -u schemas/EiffelActivityTriggeredEvent/4.2.0.json schemas/EiffelActivityTriggeredEvent/4.3.0.json
--- definitions/EiffelActivityTriggeredEvent/4.2.0.yml	2023-06-28 22:52:10.317452149 +0200
+++ definitions/EiffelActivityTriggeredEvent/4.3.0.yml	2023-06-29 15:02:32.138578544 +0200
@@ -126,6 +126,18 @@
       any_type: false
       types:
         - EiffelFlowContextDefinedEvent
+  ORIGINAL_TRIGGER:
+    description: 'Used when the current activity is a new attempt at
+      completing a previous activity, typically because the latter failed.
+      Although this activity may have been manually triggered and thus
+      lacks a cause that can be described with Eiffel, this link can be
+      used to convey the second-order cause.'
+    required: false
+    multiple: false
+    targets:
+      any_type: false
+      types:
+        - EiffelActivityTriggeredEvent
   PRECURSOR:
     description: 'Used to declare temporal relationships between
       [activities](../eiffel-syntax-and-usage/glossary.md#activity) in a
@@ -141,6 +153,8 @@
       types:
         - EiffelActivityTriggeredEvent
 _history:
+  - version: 4.3.0
+    changes: Add ORIGINAL_TRIGGER link (see [Issue 246](https://github.com/eiffel-community/eiffel/issues/246)).
   - version: 4.2.0
     introduced_in: edition-arica
     changes: Add schema URL to the meta object (see [Issue 280](https://github.com/eiffel-community/eiffel/issues/280)).
--- schemas/EiffelActivityTriggeredEvent/4.2.0.json	2023-06-24 23:13:33.814444317 +0200
+++ schemas/EiffelActivityTriggeredEvent/4.3.0.json	2023-06-29 15:02:32.138578544 +0200
@@ -18,9 +18,9 @@
         "version": {
           "type": "string",
           "enum": [
-            "4.2.0"
+            "4.3.0"
           ],
-          "default": "4.2.0"
+          "default": "4.3.0"
         },
         "time": {
           "type": "integer"
```

File: eiffel-syntax-and-usage/activity-linking.md [hunk 1]
```diff
@@ -71,6 +71,23 @@ __Legal targets:__ [EiffelActivityTriggeredEvent][ActT],
...
 
+### ORIGINAL_TRIGGER
+
+This link type is used for sequences of (usually) failing activities where the first activity declares the "root" trigger and the subsequent activities are retries of the first activity.
+
+*The image below shows how a source code change triggers a build activity. That activity fails and is retried, and the second activity (which might be manually triggered by a user) has an ORIGINAL_TRIGGER link to the first activity.*
+
+![An example of using ORIGINAL_TRIGGER](./original-trigger.png "An example of using ORIGINAL_TRIGGER")
+
+An ORIGINAL_TRIGGER link could be combined with a CAUSE link. For example, an activity that retries a failing build could conceivably have a CAUSE link to the ActF event that announces the failure, but an ORIGINAL_TRIGGER link would further increase the fidelity of the model by describing not just that the second activity is run after the first one but that the first activity is the original trigger of the activity sequence.
```
Comment (m-linner-ericsson) [inline eiffel-syntax-and-usage/activity-linking.md:82]: Had to read this paragraph more than once to understand it. Could probably be easier with a picture but if not I have made an attempt to reformulate it (but do feel free to discard it)

```suggestion
To enhance clarity, an event could combine the ORIGINAL_TRIGGER with a CAUSE link. For example: An activity that retries failing builds, based on events, might use a CAUSE link to the ActF that that announces the failure. To increase the fidelity of the model the event could also use an ORIGINAL_TRIGGER, to the ActT, to signal which event was the original trigger of the activity sequence.
```

Comment (magnusbaeck) [reply eiffel-syntax-and-usage/activity-linking.md]: Took you suggestion more or less as-in, just removing a stray word and some commas.

File: definitions/EiffelActivityTriggeredEvent/4.3.0.yml [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+      to.'
+    required: false
+    multiple: true
+    targets:
+      any_type: false
+      types:
+        - EiffelFlowContextDefinedEvent
+  ORIGINAL_TRIGGER:
+    description: 'Used when the current activity is a new attempt at
+      completing a previous activity, typically because the latter failed.
```
Comment (e-backmark-ericsson) [inline definitions/EiffelActivityTriggeredEvent/4.3.0.yml:131]: The word 'latter' can be misread here I believe, since it is sometimes used to refer to 'last in time' rather than 'last mentioned'.
```suggestion
      completing a previous activity, typically because the previous activity failed.
```

File: eiffel-syntax-and-usage/activity-linking.md [hunk 3]
```diff
@@ -71,6 +71,23 @@ __Legal targets:__ [EiffelActivityTriggeredEvent][ActT],
 [EiffelTestCaseTriggeredEvent][TCT], [EiffelTestSuiteStartedEvent][TSS]  
 __Multiple allowed:__ Yes  
 
+### ORIGINAL_TRIGGER
+
+This link type is used for sequences of (usually) failing activities where the first activity declares the "root" trigger and the subsequent activities are retries of the first activity.
+
+*The image below shows how a source code change triggers a build activity. That activity fails and is retried, and the second activity (which might be manually triggered by a user) has an ORIGINAL_TRIGGER link to the first activity.*
```
Comment (e-backmark-ericsson) [inline eiffel-syntax-and-usage/activity-linking.md:78]: It isn't necessarily a build, and that should not be important either, so let's leave that out.
```suggestion
*The image below shows how a source code change triggers an activity. That activity fails and is retried, and the second activity (which might be manually triggered by a user) has an ORIGINAL_TRIGGER link to the first activity.*
```

File: eiffel-syntax-and-usage/activity-linking.md [hunk 4]
```diff
@@ -71,6 +71,23 @@ __Legal targets:__ [EiffelActivityTriggeredEvent][ActT],
...
+
+![An example of using ORIGINAL_TRIGGER](./original-trigger.png "An example of using ORIGINAL_TRIGGER")
```
Comment (e-backmark-ericsson) [inline eiffel-syntax-and-usage/activity-linking.md:80]: Maybe the picture should show a third activity execution as well, with ORIGINAL_TRIGGER targeting the first ActT, to avoid confusion of whether a third ActT should target the first or the previous ActT? We should probably expect the readers to understand that by the phrasing in the text above though.

File: eiffel-vocabulary/EiffelActivityTriggeredEvent.md [hunk 5]
```diff
@@ -70,6 +70,12 @@ __Legal targets:__ [EiffelFlowContextDefinedEvent](../eiffel-vocabulary/EiffelFl
 __Multiple allowed:__ Yes  
 __Description:__ Identifies the flow context of the event: which is the continuous integration and delivery flow in which this occurred – e.g. which product, project, track or version this is applicable to.
 
+### ORIGINAL_TRIGGER
+__Required:__ No  
+__Legal targets:__ [EiffelActivityTriggeredEvent](../eiffel-vocabulary/EiffelActivityTriggeredEvent.md)  
+__Multiple allowed:__ No  
+__Description:__ Used when the current activity is a new attempt at completing a previous activity, typically because the latter failed. Although this activity may have been manually triggered and thus lacks a cause that can be described with Eiffel, this link can be used to convey the second-order cause.
```
Comment (e-backmark-ericsson) [inline eiffel-vocabulary/EiffelActivityTriggeredEvent.md:77]:
```suggestion
__Description:__ Used when the current activity is a new attempt at completing a previous activity, typically because the previous activity failed. Although this activity may have been manually triggered and thus lacks a cause that can be described with Eiffel, this link can be used to convey the second-order cause.
```

**Review by e-backmark-ericsson — COMMENTED** [review]
The current state of this PR is good enough, so I will not request changes to be made (in order to not hinder it being merged if I'm off later). But I have some wishes for improvements that could be considered if you agree.

File: eiffel-syntax-and-usage/activity-linking.md (hunk 3 above)
Comment (magnusbaeck) [reply eiffel-syntax-and-usage/activity-linking.md:78]: I chose to add "build" to be a bit less abstract in the example, but you're right.

File: eiffel-syntax-and-usage/activity-linking.md (hunk 4 above)
Comment (e-backmark-ericsson) [reply eiffel-syntax-and-usage/activity-linking.md:80]: Yes. Leave the picture as-is then, but maybe rephrase the text slightly. It does state "first activity", but maybe an additional sentence mentioning a third activity also linking to the first activity could be written.

File: definitions/EiffelActivityTriggeredEvent/4.3.0.yml (hunk 2 above)
Comment (magnusbaeck) [reply definitions/EiffelActivityTriggeredEvent/4.3.0.yml]: On second thoughts I'll go with "previous one" to be less repetitive. It'll remove the ambiguity just as well. I'll adjust that in a subsequent commit that adds new versions of TCT and TSS.

**Review by m-linner-ericsson — DISMISSED** [review]

File: eiffel-syntax-and-usage/activity-linking.md (hunk 4 above)
Comment (magnusbaeck) [reply eiffel-syntax-and-usage/activity-linking.md:80]: Done.

**Review by e-backmark-ericsson — APPROVED** [review]

**Review by e-backmark-ericsson — APPROVED** [review]

**Comment by magnusbaeck** [comment]
Ha, the newly added test for history table existence caught me. Sorry for not catching this before pushing this and requesting your reviews.

**Review by e-backmark-ericsson — APPROVED** [review]

**Review by m-linner-ericsson — APPROVED** [review]
